ts/modules/marker-clusters.ts

```diff
@@ -0,0 +1,1948 @@
+         * distance from the cluster center and rebuild it when the point is
+         * spaced enough to be outside the cluster. It provides performance
+         * improvement and more stable clusters position yet can be used rather
+         * on small and sparse datasets.
+         *
+         * By default, the algorithm depends on visible quantity of points
+         * and kmeansThreshold. When there are more visible points than the
+         * kmeansThreshold the `grid` algorithm is used, otherwise `kmeans`.
+         *
+         * The custom clustering algorithm can be added by assigning a callback
+         * function as the type property. This function takes an array of
+         * processedXData, processedYData, processedXData indexes and
+         * layoutAlgorithm options as arguments and should return an object
+         * with grouped data.
+         *
+         * The algorithm should return an object like that:
+         * <pre>{
+         *  clusterId1: [{
+         *      x: 573,
+         *      y: 285,
+         *      index: 1 // point index in the data array
+         *  }, {
+         *      x: 521,
+         *      y: 197,
+         *      index: 2
+         *  }],
+         *  clusterId2: [{
+         *      ...
+         *  }]
+         *  ...
+         * }</pre>
+         *
+         * `clusterId` (example above - unique id of a cluster or noise)
+         * is an array of points belonging to a cluster. If the
+         * array has only one point or fewer points than set in
+         * `cluster.minimumClusterSize` it won't be combined into a cluster.
+         *
+         * @sample maps/marker-clusters/optimized-kmeans
+         *         Optimized K-Means algorithm
+         * @sample highcharts/marker-clusters/kmeans
+         *         K-Means algorithm
+         * @sample highcharts/marker-clusters/grid
+         *         Grid algorithm
+         * @sample maps/marker-clusters/custom-alg
+         *         Custom algorithm
+         *
+         * @type {string|Function}
+         * @see [cluster.minimumClusterSize](#plotOptions.scatter.marker.cluster.minimumClusterSize)
+         * @optionparent plotOptions.scatter.marker.cluster.layoutAlgorithm.type
+         */
+        /* eslint-enable max-len */
+        /**
+         * When `type` is set to the `grid`,
+         * `gridSize` is a size of a grid square element either as a number
+         * defining pixels, or a percentage defining a percentage
+         * of the plot area width.
+         *
+         * @type    {number|string}
+         */
+        gridSize: 50,
+        /* eslint-disable max-len */
+        /**
+         * When `type` is set to `kmeans`,
+         * `iterations` are the number of iterations that this algorithm will be
+         * repeated to find clusters positions.
+         *
+         * @type    {number}
+         * @optionparent plotOptions.scatter.marker.cluster.layoutAlgorithm.iterations
```

Comment:
This makes no sense. `@optionparent` has no basic type, except an interface. Do you mean `@apioption` ?